File: src/main/java/leetcode/Problem424.java
package leetcode;

/**
 * https://leetcode.com/problems/longest-repeating-character-replacement/
 */
public class Problem424 {
    public int characterReplacement(String s, int k) {
        int max = 0;
        for (int i = 0; i < s.length(); i++) {
            int j = i;
            int[] counts = new int[26];
            int sum = 0;
            int tmpMax = 0;
            for (; j < s.length(); j++) {
                char c = s.charAt(j);
                counts[c - 'A']++;
                sum++;
                tmpMax = Math.max(tmpMax, counts[c - 'A']);
                if (sum - tmpMax > k) {
                    sum--;
                    counts[c - 'A']--;
                    break;
                }
            }
            max = Math.max(max, sum);
        }
        return max;
    }
}

File: src/main/rust/src/bin/problem424.rs
use std::collections::{HashMap, VecDeque};

// https://leetcode.com/problems/longest-repeating-character-replacement/description/
pub fn character_replacement(s: String, k: i32) -> i32 {
    let chars: Vec<char> = s.chars().collect();
    let mut answer = 0;
    let mut deque: VecDeque<usize> = VecDeque::new();
    deque.push_back(0);
    let mut map: HashMap<char, i32> = HashMap::new();
    *map.entry(chars[0]).or_insert(0) += 1;
    for i in 1..chars.len() {
        println!("{i}:");
        if chars[i - 1] != chars[i] {
            deque.push_back(i);
        }
        println!(" deque: {:?}", deque);
        *map.entry(chars[i]).or_insert(0) += 1;
        println!(" map: {:?}", map);
        while map.len() as i32 > k {
            let index = deque.pop_front().unwrap();
            let count = map.get_mut(&chars[index]).unwrap();
            *count -= 1;
            if *count == 0 {
                map.remove(&chars[index]);
            }
        }
        answer = answer.max(i as i32 - *deque.front().unwrap_or(&0) as i32 + 1);
    }
    answer
}

fn main() {
    println!("{}", character_replacement("ABAB".to_string(), 2)); // 4
    println!("{}", character_replacement("AABABBA".to_string(), 1)); // 4
    println!("{}", character_replacement("AABAABAAACAB".to_string(), 2)); // 9
    println!("{}", character_replacement("AB".to_string(), 2)); // 2
    println!("{}", character_replacement("AB".to_string(), 1)); // 2
    println!("{}", character_replacement("AABAABAAABAB".to_string(), 2)); // 9
}
